### PR TITLE
Dynamically discover lucene version for use in build

### PR DESCRIPTION
**Pull request by janhoy**
The test `:solr:validateConfigFileSanity` checks that `<luceneMatchVersion>` is correct in various solrconfig files. Now that lucene version is different from solr version (e.g. on  main branch right now), the build needs to know what lucene version we have.

I tried using Lucene's `Version.LATEST.toString()`, but the build found a `8_10_0` version in my env.
So now I try to just pull it from `versions.props`, and that works.

However, it may noe be compatible with the local lucene version override feature in `lucene-dev-repo-composite.gradle`??

### CODE REVIEW

**Comment by janhoy** [comment]
No, I have not overridden lucene version locally. But I think the gradle build may pick up another lucene-jar from .m2 than Solr itself, not sure which classpath the .gradle files use for themselves...

**Comment by dweiss** [comment]
> So now I try to just pull it from versions.props, and that works.

In short: no. If you plan to extract Lucene version/ dependency version from within the build file then you should be using getVersion("artifact coordinates") - this will work. But this method can be used at build execution time only so sometimes you have to resort to using providers. An example is in icu.gradle in Lucene:
```
  // Resolve version lazily (can't resolve at configuration time).
  def icu4jVersionProvider = project.provider { getVersion('com.ibm.icu', 'icu4j') } 
```

**Comment by dweiss** [comment]
Note: It will get a bit more complicated if you need this property at evaluation time.

> I tried using Lucene's Version.LATEST.toString(), but the build found a 8_10_0 version in my env.

I'm not sure what code you used here but if you used this exact snippet from within the build then you accessed gradle's internal references somehow. I'm surprised it even compiled (means gradle has Lucene inside somewhere?).

**Comment by janhoy** [comment]
See updated PR for a provider approach, unfortunately it also fails
```
1: Task failed with an exception.
-----------
* Where:
Script '/Users/janhoy/git/solr/gradle/testing/randomization.gradle' line: 39
* What went wrong:
A problem occurred evaluating script.
> A problem occurred configuring project ':solr:server'.
   > Not allowed to call gradle-consistent-versions's getVersion("org.apache.lucene", "lucene-core", configurations.unifiedClasspath) at configuration time
* Try:
Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output. Run with --scan to get full insights.
==============================================================================
```

**Comment by uschindler** [comment]
Naaah, that can't work as the provider is resolved a evaluation time.

**Comment by janhoy** [comment]
We can live with hardcoded '9.0.0' for now, but should find a solution to not have theTest LuceneMatchVersion test fail on you if you have a custom lucene version in dev.

**Comment by dweiss** [comment]
Right - that's why I said:

> Note: It will get a bit more complicated if you need this property at evaluation time.

The problem here is that in general you won't know the version of a dependency until the configuration is "resolved" and gradle folks try hard to defer this resolution until execution time so that only those configurations that are really needed/ used are resolved. Palantir's version consistency plugin also comes into play. It is complicated.

**Comment by dweiss** [comment]
Looking at the code I think the top-level property from build.gradle should be removed and its initialization moved entirely to lucene-dev-repo-composite.gradle - then you'd know which version is actually in use (there are hacks that read versions.props there); the only place that uses this top-level property is in randomization.gradle; this should follow lucene-dev-repo-composite.gradle in the build file inclusion order and then it'd have access to a properly evaluated Lucene version (as would solr.config-file-sanity.gradle).

**Comment by uschindler** [comment]
Another option would be to resolve the property when the test starts (e.g., using a "before" closure) and it is then added to the command line?

**Comment by uschindler** [comment]
Maybe add an JVMArgumentProvider to the test task. I can check if this works.

**Comment by uschindler** [comment]
Hi, I got it running. Will push a commit. You still may need to adopt it with the local Lucene checkout and the version number. Also the verison number may need to be made compatible to Lucene Match Version, so parsing won't fail. Maybe use a regex to extract correct parts (so it only returns "9.0").

**Comment by uschindler** [comment]
OK, see my changes. Tests are starting up correctly and it prints the lifecycle message (we should remove it). Also i checked the local "ps -ef" and have seen the command line args passed correctly.

I "normalized" the LuceneMatchVersion to be only formatted "x.y", as this is what Version class expects. Or am I wrong?

Also there are some properties passed to tests which are no longer needed. I think the `tests.LUCENE_VERSION` one is only used by a single Lucene Core test and may be obsolete. And I am not sure about `tests.bwcdir`!

**Comment by uschindler** [comment]
For some reason "solr:validateConfigFileSanity" fails on the PR. Thats unrelated I think.

**Comment by uschindler** [comment]
I think for the solr:validateConfigFileSanity test you can do the same: Just resolve the version at runtime of task. As this is a handwritten task we can easily use a provider.

Maybe make a more generic provider available in the same way like you did before and let the JVMAgrumentProvider just call it.

So solr:validateConfigFileSanity may be fixed by that. SHould I look into that or do you want to check this?

**Comment by uschindler** [comment]
Ah I see, solr:validateConfigFileSanity uses the test option. I'll fix.

**Comment by uschindler** [comment]
OK, I fixed both.

There is still one open question: Normally Lucene Version should only be provided in config files as "x.y", not "x.y.z". So I'd better go and use LuceneMatchVersion as "9.0" (so strip last part in the regex) and fix all config files.

Otherwise it will get a desaster on updates. You would need to touch config files on every bugfix release.

**Review by uschindler — APPROVED** [review]
To me this looks good now. Nevertheless I'd change a bit:
- remove the lifecycle logging or tone down to debug or info
- change Lucene Match Version to use x.y format only to ease bugfix upgrades and minor versions are irrelevant in config files.

**Comment by dweiss** [comment]
What you wrote is fine, Uwe, but it won't work with a custom Lucene location - that's why I suggested moving the entire logic you currently have in that provider to lucene-dev-repo-composite.gradle which does this (and more) to extract the locally-overridden lucene version number. I'm sorry but I can't help with this now.

**Comment by uschindler** [comment]
OK, I expected something like this. I thought that the versions are patched by the repo-composite, so it just "replaces" the dependency. But when replacing it changes its name.

I will first to setup something like this side-by-side repos in my local computer with env args (I did this a few weeks ago) to do testing and make it work with a relationship to lucene 9.x branch..

**Comment by uschindler** [comment]
> this should follow lucene-dev-repo-composite.gradle in the build file inclusion order and then it'd have access to a properly evaluated Lucene version (as would solr.config-file-sanity.gradle).

This move in order should not be needed, as the provider is only used in another provider, so the evaluation is lazy-lazy :-)

**Comment by janhoy** [comment]
This looks really good. I'm not going to test this locally myself, but I can change the luceneMatchVersion inserted into solrconfig.xml to Major.Minor. Wehenver you guys says it's ready I can then merge and backport. 

(I have hacked a cherrypick.sh script that I want to test out, streamlining backporting of trivial, non-conflicting changes) 😁

**Comment by uschindler** [comment]
> This looks really good. I'm not going to test this locally myself, but I can change the luceneMatchVersion inserted into solrconfig.xml to Major.Minor. Wehenver you guys says it's ready I can then merge and backport.

All fine. About the 9.0.0 vs. 9.0 I am not fully sure. In the past we changed this on every version, so the current state with full version is fine, although it is not needed for config file sanity (IMHO there we should only use x.y versions). When executing tests it does not matter which version we pass, we can use full version, too. It obly must be in the digit-only format, so we have to remove any appendixes.

My proposal would be: read the version as done currently and sanitize to x.y.z. Pass this to tests, but the config files shipped with release should use the x.y version to make it easier for users to understand that only this is needed for "correct config files". This check should be adapted in the config file checker. But maybe that's a separate issue, isn't it?

> (I have hacked a cherrypick.sh script that I want to test out, streamlining backporting of trivial, non-conflicting changes) 😁

I use TortoiseGit for cherrypicking. Thats grandious: Open the log window on 9.x/9.0 checkout, switch branch in log window to show main, select commit, right click, cherry-pick (also works with multiple).

**Comment by janhoy** [comment]
> I use TortoiseGit for cherrypicking. Thats grandious: Open the log window on 9.x/9.0 checkout, switch branch in log window to show main, select commit, right click, cherry-pick (also works with multiple).

I use IntelliJ and it is also very simple - git -> log -> find commit -> right-click -> cherry-pick.

But my script does a bit more, it takes branch and one or more hashes, does the cherry-picking, if not clean, open mergetool, else runs `gradlew check -x` test. So you could run one command to backport 2-3 commits to branch_9x and branch_9_0 in one go. Intended for the more trivial cases only of course...

**Comment by janhoy** [comment]
I committed the X.Y version change, and an update to `addVersion.py` tool a bit, so it does the same. Now need to fix the configfile-sanity to expect X.Y too

**Comment by uschindler** [comment]
Thanks, I can do the rest including the check script. I think I will rename the provider to luceneBaseVersionProvider.

**Comment by janhoy** [comment]
Here is what I got, which works
```patch
Index: gradle/validation/solr.config-file-sanity.gradle
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/gradle/validation/solr.config-file-sanity.gradle b/gradle/validation/solr.config-file-sanity.gradle
--- a/gradle/validation/solr.config-file-sanity.gradle	(revision 636ffd7daea7245d1b31932d549372f31b52cca5)
+++ b/gradle/validation/solr.config-file-sanity.gradle	(revision be6503578dbd8e49ac658464c3add8e28ec944c0)
@@ -21,6 +21,8 @@
   task validateConfigFileSanity() {
     doFirst {
       def matchVersion = rootProject.luceneMatchVersionProvider.get()
+      def (major, minor) = matchVersion.tokenize('.')
+      def matchVersionMajorMinor = "${major}.${minor}"
 
       // Verify solrconfig.xml files declare proper luceneMatchVersion.
       [
@@ -33,8 +35,8 @@
 
         configFiles.each { file ->
           def content = file.getText("UTF-8")
-          if (!content.contains("<luceneMatchVersion>${matchVersion}<")) {
-            throw new GradleException("Configset does not refer to the correct luceneMatchVersion (${matchVersion}): ${file.absolutePath}")
+          if (!content.contains("<luceneMatchVersion>${matchVersionMajorMinor}<")) {
+            throw new GradleException("Configset does not refer to the correct luceneMatchVersion (${matchVersionMajorMinor}): ${file.absolutePath}")
           }
         }
       }
```

**Comment by uschindler** [comment]
Oh thats also a nice one, let me change it

**Comment by uschindler** [comment]
..still fighting `lucene-dev-repo-composite.gradle`: this is the craziest thing I have ever seen!

**Comment by uschindler** [comment]
Status update:
- `-Plucene.dev.version=SOME_SNAPSHOT_FROM_LOCAL_M2_REPO` works out of box
- `-Plucene.dev.path=../lucene` (my local setup is in progress). The strange thing is that the included build is not resolved, looks like I need to add so dependency. Have to dig...

**Comment by uschindler** [comment]
Hi @dweiss I am stuck how to get the version from the included build. It looks like the included build is completely sandboxed and you only have access to dependencyResultion rules on settings phase and when running at config phase of main build the `gradle.includedBuild('lucene')` is also just a shim hiding all important facts. You can only lookup tasks, its name, and `projectDir`. The only solution I have is to do regex on build.gradle to get Lucene version.

**Comment by janhoy** [comment]
Perhaps we can pragmatically document that when using `-Plucene.dev.path=` you also need to supply `-Plucene.version=` explicitly, and make that an override that always "wins"?

**Comment by uschindler** [comment]
This is the only thing that worked, although it is a hackydihickhack hell:
```groovy
// replace Lucene version by dynamic one:
rootProject.ext.luceneBaseVersionProvider = rootProject.provider {
  (new File(gradle.includedBuild('lucene').projectDir, 'build.gradle').getText("UTF-8") =~ /baseVersion\s*=\s*'(\d+\.\d+\.\d+)'/)[0][1]
}
```

**Comment by uschindler** [comment]
I updated it with a heavy hack alarm. Basically I grep through the build.gradle file, match on the line with baseVersion and then execute the statement.

I also renamed the provider, but basically it was not complicated.

I tested 2 versions:

A `lucene/` checkout of branch_9_0 (all others are not compatible) next to my solr checkout:
```
solr$ gradlew :solr:validateConfigFileSanity -Plucene.dev.path=../lucene
Local Lucene development repository will be used substituting 9.0.0: C:\Users\Uwe Schindler\Projects\lucene\lucene

> Configure project :
Local Lucene development repository will override luceneBaseVersion with: 9.0.1

BUILD SUCCESSFUL in 3s
3 actionable tasks: 1 executed, 2 up-to-date
```

And I also deployed the 9.0.1 artifacts to local Maven:
```
lucene$ gradlew mavenToLocal
[...]
BUILD SUCCESSFUL in 2m 17s
248 actionable tasks: 177 executed, 71 up-to-date

solr$ gradlew :solr:solrj:test -Plucene.dev.version=9.0.1-SNAPSHOT
...works and ps -ef shows correct version...
```

This fails (I have a local 8.11.0 in my maven repo), as expected because the config files are wrong:
```
solr$ gradlew :solr:validateConfigFileSanity -Plucene.dev.version=8.11.0

> Configure project :
Lucene version forced by -Plucene.dev.version=8.11.0

> Task :solr:validateConfigFileSanity FAILED

FAILURE: Build failed with an exception.

* Where:
Script 'C:\Users\Uwe Schindler\Projects\lucene\solr\gradle\validation\solr.config-file-sanity.gradle' line: 38

* What went wrong:
Execution failed for task ':solr:validateConfigFileSanity'.
> Configset does not refer to the correct luceneMatchVersion (8.11): C:\Users\Uwe Schindler\Projects\lucene\solr\solr\server\solr\configsets\sample_techproducts_configs\conf\solrconfig.xml

* Try:
Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output. Run with --scan to get full insights.

* Get more help at https://help.gradle.org

BUILD FAILED in 6s
3 actionable tasks: 1 executed, 2 up-to-date
```

**Comment by dweiss** [comment]
that development build file is indeed very, very difficult because it is included in gradle at build phase - it took a great deal of time to get that working and it's indeed two different environments (as it was noted in the comments).

File: gradle/lucene-dev/lucene-dev-repo-composite.gradle [hunk 1]
```diff
@@ -130,6 +130,21 @@ if (luceneDevRepo != null) {
   } else {
     // We're being applied at build-time and Lucene development repository exists. Configure
     // certain aspects of the build so that things work with it.
+    
+    // replace luceneBaseVersionProvider by one evaluating the included build:
+    configure(rootProject) {
+      def line = new File(gradle.includedBuild('lucene').projectDir, 'build.gradle').readLines("UTF-8").find { it =~ /\bbaseVersion\s*=\s*['"]/ }
```
Comment (dweiss) [inline gradle/lucene-dev/lucene-dev-repo-composite.gradle:136]: Yep. included builds don't really have a notion of "version" - they only have coordinates (they're treated much like subprojects). I don't know if there is a better solution than that presented here.

**Review by dweiss — APPROVED** [review]

**Comment by uschindler** [comment]
> that development build file is indeed very, very difficult because it is included in gradle at build phase - it took a great deal of time to get that working and it's indeed two different environments (as it was noted in the comments).

All fine!

What I was wondering about: According to documentation the dependency substitution should be completely automatically. But it looks like we need that artifact name to project name mapping. Is this because Lucene/Solr's gradle.build files do not explicitly name "group" and "name" (those coordinates are only generated on-the-fly in the maven publishing)?

**Comment by dweiss** [comment]
> Is this because Lucene/Solr's gradle.build files do not explicitly name "group" and "name" (those coordinates are only generated on-the-fly in the maven publishing)?

Maven coordinates only exist in maven. Gradle has module paths - build inclusion is kind of hacking these two worlds together. What's in that file is a result of trial-and-error and my understanding of how these two worlds come together, I'm by no means trying to say I understand everything going on under the hood of Gradle...

**Comment by uschindler** [comment]
Thanks for merging!

**Pull request merged**